include/h2/tensor/tensor_types.hpp

```diff
@@ -0,0 +1,363 @@
+ * RawBuffer. If the Tensor is a view, only the Tensor's SyncInfo will
+ * be changed. (This is how a Tensor's SyncInfo may differ from its
+ * RawBuffer's.) This enables views of the same Tensor to enqueue
+ * operations on multiple compute streams concurrently; it is up to the
+ * user to ensure the appropriate synchronization in such uses.
+ *
+ * This requires careful handling of destruction in the RawBuffer, as
+ * there may be operations on multiple compute streams accessing the
+ * data, yet the RawBuffer is only (directly) associated with one
+ * SyncInfo. In particular, consider the case where an initial Tensor A
+ * is created with SyncInfo SA, and then a view, B, of that Tensor is
+ * created and associated with SyncInfo SB. The underlying RawBuffer
+ * will be associated with SA. If A is deleted, the RawBuffer will
+ * still exist, as B still has a reference to it. Now suppose B
+ * launches some operations on SB, then is itself deleted. The
+ * operations should continue to run fine, due to the compute stream
+ * ordering. However, the RawBuffer's deletion will be synchronized
+ * only to SA, potentially leading to a race with operations on SB. To
+ * avoid this, whenever a Tensor discards a reference to a RawBuffer,
+ * it informs the RawBuffer it is doing so, along with its current
+ * SyncInfo. If the SyncInfo differs from the RawBuffer's, it will
+ * record an event on the Tensor's SyncInfo and keep a reference to it.
+ * When the RawBuffer is deleted, it will synchronize with all recorded
+ * SyncInfos before enqueuing the delete, to avoid races.
+ *
+ * Another situation to be aware of: If you change a Tensor's SyncInfo,
+ * it is up to you to provide any needed synchronization between the
+ * original SyncInfo and the new one.
+ *
+ * An implementation note (separate from the above semantics): SyncInfo
+ * objects are stored in StridedMemory, rather than directly in a
+ * Tensor. This is just to simplify implementation.
+ */
+template <Device Dev>
+using SyncInfo = El::SyncInfo<Dev>;
+
+/** Support printing SyncInfo. */
+template <Device Dev>
+inline std::ostream& operator<<(std::ostream&, const SyncInfo<Dev>& sync);
+
+template <>
+inline std::ostream& operator<<(std::ostream& os,
+                                const SyncInfo<Device::CPU>& sync)
+{
+  os << "CPUSync";
+  return os;
+}
+
+#ifdef H2_HAS_GPU
+template <>
+inline std::ostream& operator<<(std::ostream& os,
+                                const SyncInfo<Device::GPU>& sync)
+{
+  os << "GPUSync(" << sync.Stream() << ")";
+  return os;
+}
+#endif
+
+/**
+ * Integer type used for the number of dimensions.
+ */
+using NDimType = std::int32_t;
+
+/**
+ * Integer type used for storing dimensions.
+ */
+using DimType = std::int32_t;
+
+/**
+ * Integer type used for data indices.
+ */
+using DataIndexType = std::int64_t;
+
+/**
+ * Maximum number of dimensions a tensor may have.
+ */
+static constexpr NDimType MAX_TENSOR_DIMS = 8;
```

Comment:
cuDNN defines a maximum dimension; we could grab that here when we have cuDNN, if we care that much (the current max dim is 8, so happy coincidence).